File: python/problem54.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        left = top = 0
        bottom = len(matrix) - 1
        if bottom == -1:
            return []
        right = len(matrix[0]) - 1
        if right == -1:
            return []
        result = []
        while left <= right and top <= bottom:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1
            
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1
            
            if top <= bottom:
                for i in range(right, left - 1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1
            
            if left <= right:
                for i in range(bottom, top -1, -1):
                    result.append(matrix[i][left])
                left += 1
        return result

File: Java/problem5-4.java
class Solution {
    private char[] chars;
    private int maxLength(int left, int right) {
        while (left >= 0 && right < chars.length && chars[left] == chars[right]) {
            --left;
            ++right;
        }
        return right - left - 1;
    }
    public String longestPalindrome(String s) {
        if (s.length() == 0) {
            return "";
        }
        chars = s.toCharArray();
        int middleIdx = 0, maxLen = 0;
        for (int i = 0; i < chars.length; ++i) {
            int longer = Math.max(maxLength(i, i), maxLength(i, i + 1));
            if (longer > maxLen) {
                maxLen = longer;
                middleIdx = i;
            }
        }
        int startIdx = maxLen % 2 == 0 ? middleIdx - maxLen / 2 + 1 : middleIdx - maxLen / 2;
        
        return new String(chars, startIdx, maxLen);
    }
}
